src/IO/S3/Client.cpp

```diff
@@ -91,6 +93,18 @@ bool Client::RetryStrategy::ShouldRetry(const Aws::Client::AWSError<Aws::Client:
     if (useGCSRewrite(error))
         return false;
 
+    if (auto query_context = CurrentThread::getQueryContext())
+    {
+        /// Check if query is timeout.
+        auto query_start_time = query_context->getClientInfo().initial_query_start_time;
+        auto max_execution_time = query_context->getSettingsRef()[Setting::max_execution_time].totalSeconds();
+        auto now = time(nullptr);
+        if (query_start_time > 0 && query_start_time < now
+            && max_execution_time > 0 && now > query_start_time + max_execution_time)
+            throw Exception(ErrorCodes::TIMEOUT_EXCEEDED,
```

Comment:
TBD: throw an exception or just return false?

I don't know if this affects CI.